tools/tcpaccept.py

```diff
@@ -83,9 +83,22 @@
     u8 protocol = 0;
     // workaround for reading the sk_protocol bitfield:
     bpf_probe_read(&protocol, 1, (void *)((long)&newsk->sk_wmem_queued) - 3);
-    if (protocol != IPPROTO_TCP)
-        return 0;
+    
+    // Following comments add by Joe Yin:
+    // Unfortunately,it can not work since Linux 4.10,
+    // because the sk_wmem_queued is not following the bitfield of sk_protocol.
+    // And the following member is sk_gso_max_segs.
+    // So, we can use this:
+    // bpf_probe_read(&protocol, 1, (void *)((u64)&newsk->sk_gso_max_segs) - 3);
+    // And it works from 4.10 to the latest 4.17.1, but maybe one day will fail
```

Comment:
So the tool here won't work for 4.9 and earlier, right? Can we do some dynamic check here? You can use linux `offsetof` to find member offset of a structure. By comparing two different member offsets and you may be able to accommodate both pre-4.10 and 4.10+.

Reply:
I agree that a dynamic check would be better here, rather than trying both fields. In 4.10+, the byte retrieved from `sk_wmem_queued` might actually equal `IPPROTO_TCP` by chance.

That structure is likely to be changed often, so you'll want to compare two close fields (so that the chance of something being removed between them is lower). Comparing the offsets of `__sk_flags_offset` and `sk_gso_max_segs` might do the trick.

Reply:
It can work both pre-4.10 and 4.10+. As @pchaigno said " In 4.10+, the byte retrieved from sk_wmem_queued might actually equal IPPROTO_TCP by chance." , but the probability is very low. 
As the issue "accessing bitfield for sk_protocol #270" proposaled by @brendangregg ,and the "Bitfield access #613 "  , BCC doesn't currently support accessing bitfields.
@pchaigno __sk_flags_offset is not exist at pre-4.10, so it hard to work well.

Reply:
the offsetof function is very nice.
I have modified the script. @yonghong-song